### PR TITLE
Manage duplicate patients

### PR DESCRIPTION
**Pull request by pbugni**
Review of upload from Skagit shows lots of duplicate patient rows.  This PR manages by adding a hook in the adapter class, namely if an adapter defines a `unique_key()` method, only the first match is kept.

Enhanced upsert logic to check HAPI store for matching patient; now correctly PUTs with the Patient/id if found vs POST for new, to avoid adding duplicate patients.

Workaround added for 2 digit year - set WRAP_YEAR at 2025 pending feedback in https://www.pivotaltracker.com/story/show/179168443

### CODE REVIEW

File: hydrant/adapters/sites/skagit.py [hunk 1]
```diff
@@ -46,3 +47,10 @@ def items(self):
             if value:
                 yield attr, value
 
+    def unique_key(self):
+        """Returns a key value to represent this item as unique
+
+        Defined by adapters wishing to sort out duplicates, used in
+        comparison operators.
```
Comment (ivan-c) [inline hydrant/adapters/sites/skagit.py:54]: Would defining `__eq__()` or `__hash__()` have the same effect?

File: hydrant/models/datetime.py [hunk 2]
```diff
@@ -2,13 +2,19 @@
...
+WRAP_YEAR = 2025
+
 
 def parse_datetime(data, error_subject=None, none_safe=False):
     """Parse input string to generate a UTC datetime instance
 
     NB - date must be more recent than year 1900 or a ValueError
     will be raised.  (library limitation)
 
+    WRAP_YEAR (defined above) is used to determine how a 2 digit
```
Comment (ivan-c) [inline hydrant/models/datetime.py:14]: I wonder if we can work with each site to get 4-digit years

File: hydrant/models/patient.py [hunk 3]
```diff
@@ -39,22 +49,44 @@ def as_fhir(self):
...
+
+        # FHIR spec: 'birthDate'; HAPI search: 'birthdate'
+        patient_url = (
+            f'Patient?family={self._fields["name"]["family"]}&'
+            f'given={self._fields["name"]["given"][0]}&'
+            f'birthdate={self._fields["birthDate"]}')
+
+        # Round-trip to see if this represents a new or existing Patient
+        if target_system:
+            response = requests.get('/'.join((target_system, patient_url)))
```
Comment (ivan-c) [inline hydrant/models/patient.py:73]: Having `requests` build the querystring params is slightly cleaner
```suggestion
        patient_url = "Patient"
        search_params = {
            "family": self._fields["name"]["family"],
            "given": self._fields["name"]["given"][0],
            "birthdate": self._fields["birthDate"],
        }

        # Round-trip to see if this represents a new or existing Patient
        if target_system:
            response = requests.get('/'.join((target_system, patient_url)), params=search_params)
```

**Review by ivan-c — APPROVED** [review]
Looks good!

File: hydrant/adapters/sites/skagit.py (hunk 1 above)
Comment (pbugni) [reply hydrant/adapters/sites/skagit.py:54]: yes, but this feels overreaching.   i don't want to hate myself a year down the road when i compared two Patients and ignored changes as they compared equal...

File: hydrant/models/datetime.py (hunk 2 above)
Comment (pbugni) [reply hydrant/models/datetime.py:14]: Lydia is on that exact tact.  I really don''t like this guessing / hack.  Would be better to insist on a 4 digit year format.

**Pull request merged**